src/validation.cpp

```diff
@@ -4633,14 +4642,31 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
                     std::pair<std::multimap<uint256, FlatFilePos>::iterator, std::multimap<uint256, FlatFilePos>::iterator> range = mapBlocksUnknownParent.equal_range(head);
                     while (range.first != range.second) {
                         std::multimap<uint256, FlatFilePos>::iterator it = range.first;
+                        // see if the needed child block is still in the buffer
+                        const FlatFilePos pos = it->second;
+                        bool haveChildBlock = false;
                         std::shared_ptr<CBlock> pblockrecursive = std::make_shared<CBlock>();
-                        if (ReadBlockFromDisk(*pblockrecursive, it->second, chainparams.GetConsensus()))
-                        {
+                        // performance optimization -- else path would always work
+                        if (dbp && pos.nFile == dbp->nFile && blkdat.SetPos(pos.nPos - sizeof(uint32_t))) {
+                            // block is still in memory, positioned to the start of the size field
+                            haveChildBlock = true;
+                            blkdat.SetLimit(pos.nPos);
+                            blkdat >> nSize;
+                            CBlock& blockrecursive = *pblockrecursive;
+                            blkdat.SetLimit(pos.nPos + nSize);
+                            blkdat >> blockrecursive;
+                            LogPrint(BCLog::REINDEX, "%s: Processing in-memory out of order child %s of %s\n", __func__, pblockrecursive->GetHash().ToString(),
+                                    head.ToString());
```

Comment:
I don't know how to write a unit test for this PR (there are no existing unit tests for any of this area of the code), but `test/functional/wallet_basic.py` does run through the new code -- you can add an `assert(false)` here and that test fails, so that's at least something.